### PR TITLE
kernel: Stub out device pass through support

### PR DESCRIPTION
**Pull request by alistair23**
### Pull Request Overview

This is an implementation of device pass through support, as mentioned in https://github.com/tock/tock/issues/4020

This creates the Rust traits to implement device pass through to a userspace application.

This is disabled by default in the kernel. A board and a chip implementation must explicitly opt into it

### Testing Strategy

I have a BLE client up and running in userspace using this. I can connect to a Tock board and read Gatt data, with the entire BLE stack running in userspace.

### TODO or Help Wanted

This feels ready to review. If this is accepted I will start to send PRs for the implementations

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make prepush`.

### CODE REVIEW

File: kernel/src/platform/platform.rs
Comment (alevy) [inline kernel/src/platform/platform.rs:49]:
```suggestion
    type DevicePassthroughFilter: DevivePassthroughFilter;
```

**Review by lschuermann — CHANGES_REQUESTED** [review]
I'm generally in favor of this mechanism. While I think that the basic intuition behind this interface makes sense, I don't agree with this particular proposal: in short, I think that `allocate_app_device_region` is redundant with the existing `allocate_region`, and `passthrough_filter` will be redundant with the existing `process_fault_hook` for virtually all peripherals.

"Passthrough" of devices requires us to, at least:
- provide a userspace app access to some slice of MMIO memory, and
- handle other privileged system interactions, such as forwarding interrupts to userspace.

What you propose in this PR (as I understand it) is an interface quite similar to one used for TLB refills: for a given access (denoted by `addr` and `size`), return the largest matching contiguous region of memory that should be mapped / made accessible. This makes sense.

However, when we consider the second bullet above, it becomes clear that for virtually all peripherals we will want to have some kernel component responsible for bridging privileged system interactions into userspace (e.g., converting interrupts into upcalls), which I will call a _shim_. I imagine this as a privileged capsule which implements the `SyscallDriver` trait.

With this shim though, we have established an entity in the system that has a (limited) understanding of the particular semantics of the specific peripheral we're exposing to userspace: for instance, whether it is currently safe to expose access to it, whether the peripheral requires some de/re-initialization across process faults, or time-sharing a peripheral between processes. As such, this component will be particularly well suited to govern when and whether a process should have access to a particular part of an MMIO peripheral's address space.

We also do not need another core-kernel trait to build this component today---the `ProcessFault` handler is exactly the right abstraction for it. A process performing MMIO accesses to a peripheral (perhaps after it has claimed ownership of the peripheral through a system call) will cause a fault, similar to your proposal. A board can then register the peripheral's shim as a handler for these faults. The shim, being a privileged component, can add or remove the MMIO region from the app's allowed MPU regions.

This design preserves the rather particular nature of this abstraction: the shim retains a high-degree of control over the peripheral and can reason about the precise safety implications of providing access to (a part of) the peripheral's register file. You could even imagine excluding, e.g., DMA registers from the address space and emulating them in the kernel.

Finally, I don't understand the benefit of adding a dedicated `allocate_app_device_region` method. It's concerns around safety, concurrent accesses and aliasing, etc. seem virtually identical to those we (ought to) have on the regular `allocate_region` function.

Ultimately, what would help here is to have some small end-to-end example of a peripheral exposed to userspace. It's hard to reason about this mechanism in isolation, and even if we have one example I'm somewhat strongly opposed to promoting it to a general mechanism. I believe that we have all the tools necessary to build this in the kernel already, it should be straightforward to encapsulate this into some dedicated per-peripheral component.

File: kernel/src/platform/platform.rs
Comment (alistair23) [reply kernel/src/platform/platform.rs:81]: Fine with me. Is that what everyone wants?

**Comment by lschuermann** [comment]
> I don't think we should time share between processes!

I agree that this is not a feature we should seek to support in this initial proposal. However, there's another form of peripheral sharing happening between processes even if we don't time-share: when a process faults or restarts, it should be able to re-gain access to the peripheral in question. By dedicating a kernel shim (what I believe you call a "system capsule") to manage this device and it's MMIO regions, we can perform the necessary cleanup and reset as soon as an access fault comes in from the new, restarted process. Semantically, this is not much different from moving a peripheral from one process to another.

> I personally feel adding a new one makes it clearer when and how this is being used, instead of hiding it in ProcessFault.

I can see that, but in this particular case I'd argue for using `ProcessFault`, in particular as this should not be a general mechanism for now. `ProcessFault` can be used to handle a strict superset of use-cases and if you look into that type's documentation (that I believe you introduced?) you'll see documented use-cases that are not dissimilar from what this is trying to do.

The `ProcessFault` infrastructure will also not need us to commit to having the fault-then-allow MPU configuration approach. The kernel shim can, for instance, just set up the mapping upon a command system call. I'm wary of introducing dedicated infrastructure for something that's somewhat of a fringe usecase that we don't (yet) understand all implications of, and where we already have infrastructure in place that would be suitable.

> The implementation will be the same though. You can't fault on each MMIO access as that's too slow. So on the first fault you will just allocate the entire region.
> 
> The MPU granularity is generally too large, so you end up over allocating the whole region anyway.

I'm not arguing for doing anything different than what this PR, in essence, proposes. I'm not opposed to allocating a region on first use.

> The existing ones check against app memory

I'm not sure I follow. I'm not talking about `MPU::allocate_app_memory_region`, but `MPU::allocate_region`. The latter seems virtually equivalent to what you're proposing.

> It also sets us up to extend this to support the IOPMP in the future

Given that we don't have a chip with IOPMP there, I'd say let's cross that bridge when we get there?

**Comment by lschuermann** [comment]
An aside: I'm not opposed to allocating an MPU region for a peripheral on first use, but I do think it has its drawbacks. Namely, if we, for some reason, cannot give the app access to the peripheral (for example, because we've exhausted all available MPU regions), then there is no way that we can recover from this condition within an application -- the kernel would need to let it fault.

If we instead request access to a peripheral through a command system call first, then we can return an appropriate error if this is disallowed (e.g., as a result of MPU configuration failure, an app not having the right permissions, etc.). I'd be in favor of this more graceful mechanism. This interface can be exposed as part of the kernel shim driver.

**Comment by alistair23** [comment]
> 1. We should use the existing `MPU::allocate_region` instead of introducing `allocate_app_device_region`, and punt on a dedicated interface for IOPMP until we have support for that.

I don't think that's the way to go. The documentation for `allocate_region()` specifically talks about "within the specified stretch of unallocated memory". We are not trying to allocate anything in unallocated meory, so I think it's misleading to re-use the function.

Obviously we can change the docs, but that seems more invasive then adding a new function. 

> 2\. Let's build a dedicated "system capsule" / "kernel shim" that is responsible for managing the peripheral's MPU regions and interrupts. We may want to write guidelines for these capsules / shims.

That exists, it's the next PR

You can see the current example here: https://github.com/alistair23/tock/blob/alistair/device-pass-thru/kernel/src/passthrough.rs

> The `ProcessFault` infrastructure will also not need us to commit to having the fault-then-allow MPU configuration approach. The kernel shim can, for instance, just set up the mapping upon a command system call. I'm wary of introducing dedicated infrastructure for something that's somewhat of a fringe usecase that we don't (yet) understand all implications of, and where we already have infrastructure in place that would be suitable.

I don't follow this part. If we use the `ProcessFault` how can we use anything other then a fault-then-allow configuration?

If an application requests a mapping via a system call how do we know if we should/shouldn't allow it?

> If we instead request access to a peripheral through a command system call first, then we can return an appropriate error if this is disallowed (e.g., as a result of MPU configuration failure, an app not having the right permissions, etc.). I'd be in favor of this more graceful mechanism. This interface can be exposed as part of the kernel shim driver that also routes upcalls.

> If we instead request access to a peripheral through a command system call first, then we can return an appropriate error if this is disallowed (e.g., as a result of MPU configuration failure, an app not having the right permissions, etc.). I'd be in favor of this more graceful mechanism. This interface can be exposed as part of the kernel shim driver that also routes upcalls.

Agreed. That's the approach I'm taking here. An application can request the pass through and it either succeeds or fails.

I don't follow how using `ProcessFault` allows this

**Comment by alistair23** [comment]
I guess to summarise what I think the different approaches are (let me know if this is incorrect).

@lschuermann you want to achieve a similar goal with minimal changes. So basically re-use the existing `ProcessFault` and `allocate_region()`. It would still require a change to pass interrupts back to userspace, but otherwise re-use existing infrastructure. The advantage being that we don't need to add anything specific, which kind of indicates this isn't something that should be use lightly.

My approach (this PR currently) is instead adding a few new functions to explicitly enable/allow this. The idea being that it's obvious when this is allowed and in use.

What do others think?

**Comment by alistair23** [comment]
Ok. Does someone mind explaining how we can use the `ProcessFault` with a syscall to request access? That's the part I'm missing

**Comment by lschuermann** [comment]
> Does someone mind explaining how we can use the `ProcessFault` with a syscall to request access?

I didn't mean to say that you would use those together. There seem to be three basic ways to implement raw MMIO access in userspace:

1. Have userspace issue a command system call to a privileged shim driver, which then adds a region to its MPU config object.
2. Have userspace simply access the device's MMIO space, which causes a fault caught by `ProcessFault`. This can then emulate the memory access and return to userspace. This would likely be too slow to be practical.
3. Same as (2), but on fault add the MMIO region lazily (effectively mimicking a TLB refill).

Out of of the above, both (1) and (3) seem reasonable. On the one hand, (1) has the advantage of being able to return an error when the device is not accessible / the process is out of MPU regions, etc. On the other hand, (3) would transparently work for unmodified applications.

I'd prefer (1) over (3), as it makes failures explicit. @alistair23 for (1), am I correct that we wouldn't need any handler to catch faults?

**Comment by alistair23** [comment]
> 1. Have userspace issue a command system call to a privileged shim driver, which then adds a region to its MPU config object.

That's this approach. We don't need any handler to catch faults.

> 2. Have userspace simply access the device's MMIO space, which causes a fault caught by `ProcessFault`. This can then emulate the memory access and return to userspace. This would likely be too slow to be practical.

Agreed, that will be too slow

>     3. Same as (2), but on fault add the MMIO region lazily (effectively mimicking a TLB refill).

This is the `ProcessFault` approach. I don't think this is the way to go, but I thought you were advocating for it.

Although nothing stops this being build on top of this approach

**Comment by alistair23** [comment]
I'm a little confused, but it sounds like @lschuermann doesn't think this should use `ProcessFault`

So this PR needs to be updated to:

> We should use the existing MPU::allocate_region instead of introducing allocate_app_device_region, and punt on a dedicated interface for IOPMP until we have support for that.

But otherwise everyone is happy?

**Comment by lschuermann** [comment]
I think we're getting down to the roots of my confusion -- if we allocate MPU regions ahead of time, through a command system call, what's the purpose of your proposed `DevicePassthroughFilter` then? When is `filter_passthrough` called?

My interpretation of this was that it's called on fault, quite similar to when the `ProcessFault` is invoked?

**Comment by alistair23** [comment]
> I think we're getting down to the roots of my confusion -- if we allocate MPU regions ahead of time, through a command system call, what's the purpose of your proposed `DevicePassthroughFilter` then? When is `filter_passthrough` called?
> 
> My interpretation of this was that it's called on fault, quite similar to when the `ProcessFault` is invoked?

The userspace process triggers a syscall to allocate a region, to a special kernel capsule (like IPC). This capsule then checks with the board via the `DevicePassthroughFilter` to determine if the request should be allowed or not. `DevicePassthroughFilter` is just to approve of deny a MPU/PMP allocation

**Comment by lschuermann** [comment]
> The userspace process triggers a syscall to allocate a region, to a special kernel capsule (like IPC). This capsule then checks with the board via the `DevicePassthroughFilter` to determine if the request should be allowed or not. `DevicePassthroughFilter` is just to approve of deny a MPU/PMP allocation

Okay, this clears up the misunderstanding on my end! I was under the impression that this interface would be used to dynamically add MPU regions in response to a fault, which is why I thought it to be redundant with `FaultPolicy`.

I'm still not convinced that this needs to be in our platform struct. If we have a privileged capsule that controls these MMIO-accesses, the capsule itself can simply be provided a function reference that it invokes. That avoid polluting global platform state with information only related to one capsule.

Furthermore, as discussed previously, I think that we should -- for now -- have one such capsule per (group of) peripherals that we want to pass to userspace, and avoid making this a generic piece of infrastructure. This allows us to more carefully reason about the safety implications on a case-by-case basis. It also avoids trying to build a generic piece of infrastructure that needs to understand various other aspects of device behavior, such as interrupts or MMIO.

So my concrete suggestion is to remove `filter_passthrough` and instead introduce a capsule that is specialized to the exact peripheral you're interested in enabling userspace access for. This can be essentially a regular capsule, with an unsafe constructor and a special reference to the kernel to add MPU regions to an app (likely guarded by a capability such as `ProcessManagementCapability`).

**Comment by alistair23** [comment]
I have pushed an update that should address some of your concerns. There is no more modifications to the platform struct, which makes the whole feature less invasive.

I don't think we should have a capsule per peripheral. That feels clunky and I think we will end up with lots of very similar capsules duplicated everywhere. It also gives the impression that those peripherals are "safe", which is not the case

**Comment by lschuermann** [comment]
> I have pushed an update that should address some of your concerns. There is no more modifications to the platform struct, which makes the whole feature less invasive.

Thanks, I think that's better. Ultimately I don't see a need for the `DevicePassthroughFilter` to live `platform.rs` at all; it can just be defined in the same module that will ultimately define the capsule that exposes this peripheral.

> I don't think we should have a capsule per peripheral. That feels clunky and I think we will end up with lots of very similar capsules duplicated everywhere. It also gives the impression that those peripherals are "safe", which is not the case

Gotcha. I think this PR has reached a stage where we're disagreeing less on technical details and more on the high-level philosophy of how we want to support this use-case. I think that this is a good thing to discuss in a larger group. I'd propose to put this on the agenda of the next core team call (Friday 19th)?

**Comment by alevy** [comment]
I generally tend to agree we don't need a capsule per peripheral. I can imagine at least two ways this could be used:

1. As @alistair23 suggests, through a dedicated passthrough capsule that the process invokes a syscall to
2. The process loader can surmize the request from a TBF header and pre-populate the MPU on process load

In either case, this is a generic way for processes to _request_ access for that single mechanism to query permission through the trait described here.

**Comment by lschuermann** [comment]
@alevy My argument for why we want to have a capsule per peripheral is that for virtually all peripherals, MMIO is access simply won't be enough. At the very least, these peripherals will also want to have one or more interrupts to be routed into upcalls, which are in turn delivered to the app that also has MMIO access. This is not even covering peripherals which are more deeply integrated into the overall system, such as ones that integrate with DMA engines which in turn are exposed via separate MMIO regions, interrupts, and possibly even shared with other system components.

I think that a capsule is the exact right abstraction to collect and appropriately model all of these aspects of a given peripheral, and provide userspace with a neat way to access them under a single endpoint. The existence of a capsule also means that someone has already worked on bridging a given peripheral into userspace, so there's a reasonable expectation of it working. I agree with @alistair23 that the perception of safety that stems from that (even with a dedicated capsule, it might still break Tock's threat model), but I believe that this can be solved by making that capsule unsafe to construct, and with appropriate documentation.

That being said, this is just my opinion. While I'm most careful here to ensure that as a project we position ourselves in a way that doesn't make false promises around safety & security (and either way, this mechanism won't be safe _in general_), I acknowledge there are multiple valid ways to implement this. I have a preference for the above, but I'm happy to be convinced of this more general proposal -- hence requesting this to be discussed with a wider group of people.

 Arguments going for this proposal are that it's probably lower friction to support a new peripheral (although in this case I'd want to see interrupt delivery be solved in a similar generic and discoverable way). I still see an advantage in reasoning about this for each individual peripheral, and codifying the result of that thought process. That doesn't _need_ to be in the form of a capsule per peripheral, but is the closest approach to that goal that I can think of.

@alevy @alistair23 Let me know if this makes sense. I'm happy to rephrase or elaborate more.

**Comment by alistair23** [comment]
> Gotcha. I think this PR has reached a stage where we're disagreeing less on technical details and more on the high-level philosophy of how we want to support this use-case. I think that this is a good thing to discuss in a larger group. I'd propose to put this on the agenda of the next core team call (Friday 19th)?

I won't be there though, so it's just other people talking about the thing that I'm trying to implement, which doesn't feel great.

> That being said, this is just my opinion. While I'm most careful here to ensure that as a project we position ourselves in a way that doesn't make false promises around safety & security (and either way, this mechanism won't be safe _in general_), I acknowledge there are multiple valid ways to implement this. I have a preference for the above, but I'm happy to be convinced of this more general proposal -- hence requesting this to be discussed with a wider group of people.

I have follow up PRs that build on this and support exposing BLE to userspace.

I think device pass through is tricky to get right, especially as it's inherently unsafe. One option might be to continue down this path as it doesn't seem to greatly offend anyone and then re-evaluate at the end. There isn't really anything here that prevents a capsule being written per peripheral. So we can always adapt

**Comment by alevy** [comment]
Just a note that this is backwards. High-level agreement on how we want to support a use case should really come before disagreeing/discussing technical implementation details.

> I think this PR has reached a stage where we're disagreeing less on technical details and more on the high-level philosophy of how we want to support this use-case.

**Comment by alevy** [comment]
I similarly don't see an issue with this current PR being used in the way Leon describes (though I still am inclined towards what I believe is Alistair's vision in this case)

Importantly, this is obviously required that processes with such access are both highly trusted and also tightly coupled with the hardware (really, I think it might be reasonable to think of them as part of the kernel from a trust or safety perspective). So it's kind of ok for this new interface to be extremely not stable IMO if further exploration determines it should adjust.

**Comment by alistair23** [comment]
> really, I think it might be reasonable to think of them as part of the kernel from a trust or safety perspective

That's my thinking as well

File: kernel/src/platform/platform.rs
Comment (bradjc) [inline kernel/src/platform/platform.rs:268]:
```suggestion
```

File: kernel/src/platform/platform.rs
Comment (bradjc) [inline kernel/src/platform/platform.rs:234]: Devive?

**Comment by lschuermann** [comment]
> High-level agreement on how we want to support a use case should really come before disagreeing/discussing technical implementation details.

> Importantly, this is obviously required that processes with such access are both highly trusted and also tightly coupled with the hardware (really, I think it might be reasonable to think of them as part of the kernel from a trust or safety perspective). So it's kind of ok for this new interface to be extremely not stable IMO if further exploration determines it should adjust.

I agree. In this case, would it be more productive to try and determine the exact safety guarantees and invariants before we proceed with merging any technical implementation, and how this subsystem would / should fit into the kernel more generally? I think that either an issue (e.g., #4020) or a synchronous discussion with parties interested could be more productive than this PR. My focus on the technical implementation stems from the fact that this PR as a medium proposes a given implementation and solicits feedback for that.

> I won't be there though, so it's just other people talking about the thing that I'm trying to implement, which doesn't feel great.

Yes, I agree that we should avoid that. I would be more than happy to talk at a time convenient for you.

**Comment by alevy** [comment]
A design statement (perhaps just an articulation in an issue replicated in the module comment or something) would be good. If/when stabilized this could be moved to somewhere more global and formal IMO.

> the exact safety guarantees and invariants before we proceed with merging any technical implementation

I think the right invariants should be that a process with access to _any_ device pass through is totally trusted---it is part of the TCB. There are protections in place still, but those are doing providing least-privilege, _not_ strict isolation.

Specifically, any direct access to hardware may in some way allow a process to convince the device to read/write arbitrary memory in practice (whether through direct DMA access or some other series of obscure hardware operations). Capsules that have direct memory access to DMA registers, on the other hand, do not have _unrestricted_ access to those registers---the type system is meant to restrict that they can only pass valid memory they have access to. No such protections exist here, so a process can essentially do whatever it wants.

A process written in Rust (or some similarly type safe language) _could_ have a smaller TCB by essentially pulling the same trick as capsules in the kernel and having a thin wrapper around the device which is trusted, but that's beyond the scope of this subsystem design, except for the fact that it should technically be possible (like you could _probably_ run a Tock kernel inside a Tock process).

It is _possible_ that an IOMPU could help curtail such a process, but that's very TBD (I'm interested! but just no promises).

**Comment by alistair23** [comment]
What's the status here?

**Comment by lschuermann** [comment]
@alistair23 We talked about this on a high level on the core WG call of 19th July; here are the notes: https://github.com/tock/tock/blob/master/doc/wg/core/notes/core-notes-2024-07-19.md#device-passthrough

To summarize:
- Generally, we think that this is useful and solves real problems (e.g., support BLE, which seems to virtually always require running blobs)
- We cannot reason about the safety implications of doing this for arbitrary peripherals
- There are concerns about the generality of this interface. We want to codify, in some way, that we reasoned about the fact that this is reasonably safe for a given peripheral + app combination.

Because it is hard to reason about the safety impacts of this proposal in the general case, a sentiment that's shared by some people on the call was that a narrower interface that works for specific peripherals and does not touch the core kernel may need to meet a lower bar than a more general interface.

Another point that was raised is that this change should be accompanied by some abstract documentation (e.g., RFC or TRD). As part of that process, we should formalize how this interface works and its interactions and implications with Tock and its threat model.

Apart from these general comments, we refrained from talking over the specifics of this PR as we'd like you to be present for those discussions. That being said, it does seem like a synchronous high-bandwidth discussion that talks about the approach and architecture that is being proposed (and how it compares to alternative designs) seems like a reasonable next step. If that sounds good to you, we should schedule a time that works well for you.

**Comment by alistair23** [comment]
I'm going to close this and replace it with https://github.com/tock/tock/pull/4168 which includes some extra documentation and the implementation

> Because it is hard to reason about the safety impacts of this proposal in the general case, a sentiment that's shared by some people on the call was that a narrower interface that works for specific peripherals and does not touch the core kernel may need to meet a lower bar than a more general interface.

I do think that's something that we can't do. We can't reason that this hardware is safe, at least not for anything complex (like BLE). I think we should just say it's "up to the board and you are on your own" instead of trying to verify that things are safe.

I like Amit's description here: https://github.com/tock/tock/pull/4044#issuecomment-2227649303

> obviously required that processes with such access are both highly trusted and also tightly coupled with the hardware (really, I think it might be reasonable to think of them as part of the kernel from a trust or safety perspective)

**Pull request closed without merging**
